File: bubbler_app/lib/data/repositories/user_repository.dart
import '../../core/api/api_client.dart';
import '../../core/api/api_exception.dart';
import '../../core/api/endpoints.dart';
import '../models/graph.dart';
import '../models/interaction.dart';
import '../models/post.dart';
import '../models/user.dart';

/// Profile, posts, interactions, likes, and account deletion.
class UserRepository {
  UserRepository(this._client);

  final ApiClient _client;

  /// Own profile (`GET /user/me/profile`).
  Future<User> getProfile() async {
    final data = await _client.authorizedRequest(
      path: Endpoints.userMeProfile,
    );
    return User.fromJson(_asJsonMap(data));
  }

  /// Public profile by username (`GET /user/{username}/profile`).
  Future<User> getUser(String username) async {
    final data = await _client.authorizedRequest(
      path: Endpoints.userProfile(username),
    );
    return User.fromJson(_asJsonMap(data));
  }

  /// Updates email (`PUT /user/me/profile/email`).
  Future<User> updateEmail(String email) async {
    final data = await _client.authorizedRequest(
      path: Endpoints.userMeProfileEmail,
      method: 'PUT',
      data: {'email': email},
      contentType: 'application/json',
    );
    return User.fromJson(_asJsonMap(data));
  }

  /// Updates password (`PUT /user/me/profile/password`).
  Future<void> updatePassword({
    required String emailOrUsername,
    required String currentPassword,
    required String newPassword,
    required String confirmNewPassword,
  }) async {
    await _client.authorizedRequest(
      path: Endpoints.userMeProfilePassword,
      method: 'PUT',
      data: {
        'email_or_username': emailOrUsername,
        'current_password': currentPassword,
        'new_password': newPassword,
        'confirm_new_password': confirmNewPassword,
      },
      contentType: 'application/json',
    );
  }

  /// Own posts (`GET /user/me/posts`).
  Future<List<Post>> getMyPosts() async {
    final data = await _client.authorizedRequest(
      path: Endpoints.userMePosts,
    );
    return _decodePosts(data);
  }

  /// Public posts for [username] (`GET /user/{username}/posts`).
  Future<List<Post>> getUserPosts(String username) async {
    final data = await _client.authorizedRequest(
      path: Endpoints.userPosts(username),
    );
    return _decodePosts(data);
  }

  /// Recent interactions for the Bubble Trail (`GET /user/me`, capped at 20).
  Future<List<Interaction>> getMyInteractions() async {
    final data = await _client.authorizedRequest(
      path: Endpoints.userMe,
    );
    return _decodeInteractions(data);
  }

  /// All liked post IDs for the current user (`GET /user/me/likes`).
  Future<List<String>> getLikedPostIds() async {
    final data = await _client.authorizedRequest(
      path: Endpoints.userMeLikes,
    );
    if (data is! List) {
      throw const ApiInvalidResponse();
    }
    return data.map((e) => e as String).toList();
  }

  /// Records a graph interaction (`POST /user/me/interactions`).
  Future<void> recordInteraction(GraphInteractionPayload payload) async {
    await _client.authorizedRequest(
      path: Endpoints.userMeInteractions,
      method: 'POST',
      data: payload.toJson(),
      contentType: 'application/json',
    );
  }

  /// Removes a like (`DELETE /user/me/interactions/{postId}/like`).
  Future<void> deleteLike(String postId) async {
    await _client.authorizedRequest(
      path: Endpoints.userMeInteractionLike(postId),
      method: 'DELETE',
    );
  }

  /// Deletes the signed-in account (`DELETE /user/me`).
  Future<void> deleteAccount() async {
    await _client.authorizedRequest(
      path: Endpoints.userMe,
      method: 'DELETE',
    );
  }

  static List<Post> _decodePosts(dynamic data) {
    if (data is! List) {
      throw const ApiInvalidResponse();
    }
    return data
        .map((e) => Post.fromJson(_asJsonMap(e)))
        .toList();
  }

  static List<Interaction> _decodeInteractions(dynamic data) {
    if (data is! List) {
      throw const ApiInvalidResponse();
    }
    return data
        .map((e) => Interaction.fromJson(_asJsonMap(e)))
        .toList();
  }

  static Map<String, dynamic> _asJsonMap(dynamic data) {
    if (data is Map<String, dynamic>) {
      return data;
    }
    if (data is Map) {
      return Map<String, dynamic>.from(data);
    }
    throw const ApiInvalidResponse();
  }
}

File: BubblerAndroid/app/src/main/java/com/bubbler/android/data/repository/UserRepository.kt
package com.bubbler.android.data.repository

import com.bubbler.android.core.auth.TokenStore
import com.bubbler.android.core.network.ApiClient
import com.bubbler.android.core.network.Endpoints
import com.bubbler.android.core.storage.DataExportWriter
import com.bubbler.android.data.model.GraphInteractionPayload
import com.bubbler.android.data.model.Interaction
import com.bubbler.android.data.model.Post
import com.bubbler.android.data.model.User
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import okhttp3.RequestBody.Companion.toRequestBody

/**
 * Profile, posts, interactions, likes, account mutations.
 * (Email / password / delete account from iOS APIClient user routes.)
 */
open class UserRepository(
    private val apiClient: ApiClient,
    private val tokenStore: TokenStore,
) {
    open suspend fun getProfile(): User =
        apiClient.get(Endpoints.USER_ME_PROFILE, token = tokenStore.requireAccessToken())

    open suspend fun getUser(username: String): User =
        apiClient.get(Endpoints.userProfile(username), token = tokenStore.requireAccessToken())

    open suspend fun updateEmail(email: String): User {
        val payload = apiClient.json.encodeToString(
            EmailUpdateBody.serializer(),
            EmailUpdateBody(email = email),
        )
        return apiClient.request(
            path = Endpoints.USER_ME_PROFILE_EMAIL,
            method = "PUT",
            token = tokenStore.requireAccessToken(),
            body = payload.toRequestBody(ApiClient.JSON_MEDIA_TYPE),
            contentType = "application/json",
        )
    }

    open suspend fun updatePassword(
        emailOrUsername: String,
        currentPassword: String,
        newPassword: String,
        confirmNewPassword: String,
    ) {
        val payload = apiClient.json.encodeToString(
            PasswordUpdateBody.serializer(),
            PasswordUpdateBody(
                emailOrUsername = emailOrUsername,
                currentPassword = currentPassword,
                newPassword = newPassword,
                confirmNewPassword = confirmNewPassword,
            ),
        )
        apiClient.executeIgnoringBody(
            path = Endpoints.USER_ME_PROFILE_PASSWORD,
            method = "PUT",
            token = tokenStore.requireAccessToken(),
            body = payload.toRequestBody(ApiClient.JSON_MEDIA_TYPE),
            contentType = "application/json",
        )
    }

    open suspend fun getMyPosts(): List<Post> =
        apiClient.get(Endpoints.USER_ME_POSTS, token = tokenStore.requireAccessToken())

    open suspend fun getUserPosts(username: String): List<Post> =
        apiClient.get(Endpoints.userPosts(username), token = tokenStore.requireAccessToken())

    /** Recent interactions for the profile Bubble Trail (server caps at 20). */
    open suspend fun getMyInteractions(): List<Interaction> =
        apiClient.get(Endpoints.USER_ME, token = tokenStore.requireAccessToken())

    /** All post IDs the current user has liked (uncapped). */
    open suspend fun getLikedPostIds(): List<String> =
        apiClient.get(Endpoints.USER_ME_LIKES, token = tokenStore.requireAccessToken())

    open suspend fun recordInteraction(payload: GraphInteractionPayload) {
        val body = apiClient.json.encodeToString(
            GraphInteractionPayload.serializer(),
            payload,
        )
        apiClient.executeIgnoringBody(
            path = Endpoints.USER_ME_INTERACTIONS,
            method = "POST",
            token = tokenStore.requireAccessToken(),
            body = body.toRequestBody(ApiClient.JSON_MEDIA_TYPE),
            contentType = "application/json",
        )
    }

    open suspend fun deleteLike(postId: String) {
        apiClient.executeIgnoringBody(
            path = Endpoints.userInteractionLike(postId),
            method = "DELETE",
            token = tokenStore.requireAccessToken(),
        )
    }

    open suspend fun deleteAccount() {
        apiClient.executeIgnoringBody(
            path = Endpoints.USER_ME,
            method = "DELETE",
            token = tokenStore.requireAccessToken(),
        )
    }

    /**
     * Fetches the account export as pretty-printed JSON bytes ready for SAF save.
     */
    open suspend fun exportUserData(): ByteArray {
        val raw = apiClient.execute(
            path = Endpoints.USER_ME_EXPORT,
            method = "GET",
            token = tokenStore.requireAccessToken(),
        )
        return DataExportWriter.prettyPrintJson(raw)
    }
}

@Serializable
private data class EmailUpdateBody(
    val email: String,
)

@Serializable
private data class PasswordUpdateBody(
    @SerialName("email_or_username")
    val emailOrUsername: String,
    @SerialName("current_password")
    val currentPassword: String,
    @SerialName("new_password")
    val newPassword: String,
    @SerialName("confirm_new_password")
    val confirmNewPassword: String,
)
